runtime/caml/mlvalues.h

```diff
@@ -98,47 +98,49 @@ bits  31    10 9     8 7   0
 
 For 64-bit architectures:
 
-     +--------+-------+-----+
-     | wosize | color | tag |
-     +--------+-------+-----+
-bits  63    10 9     8 7   0
+     +----------+--------+-------+-----+
+     | reserved | wosize | color | tag |
+     +----------+--------+-------+-----+
+bits  63    64-R 63-R  10 9     8 7   0
 
-For x86-64 with Spacetime profiling:
-  P = PROFINFO_WIDTH (as set by "configure", currently 26 bits, giving a
-    maximum block size of just under 4Gb)
-     +----------------+----------------+-------------+
-     | profiling info | wosize         | color | tag |
-     +----------------+----------------+-------------+
-bits  63        (64-P) (63-P)        10 9     8 7   0
+where 0 <= R <= 31 is HEADER_RESERVED_BITS, set with the
+--enable-reserved-header-bits=R argument to configure.
 
 */
 
-#define Tag_hd(hd) ((tag_t) ((hd) & 0xFF))
-
-#define Gen_profinfo_shift(width) (64 - (width))
-#define Gen_profinfo_mask(width) ((1ull << (width)) - 1ull)
-#define Gen_profinfo_hd(width, hd) \
-  (((mlsize_t) ((hd) >> (Gen_profinfo_shift(width)))) \
-   & (Gen_profinfo_mask(width)))
-
-#ifdef WITH_PROFINFO
-#define PROFINFO_SHIFT (Gen_profinfo_shift(PROFINFO_WIDTH))
-#define PROFINFO_MASK (Gen_profinfo_mask(PROFINFO_WIDTH))
-/* Use NO_PROFINFO to debug problems with profinfo macros */
-#define NO_PROFINFO 0xff
-#define Hd_no_profinfo(hd) ((hd) & ~(PROFINFO_MASK << PROFINFO_SHIFT))
-#define Wosize_hd(hd) ((mlsize_t) ((Hd_no_profinfo(hd)) >> 10))
-#define Profinfo_hd(hd) (Gen_profinfo_hd(PROFINFO_WIDTH, hd))
-#else
-#define NO_PROFINFO 0
-#define Wosize_hd(hd) ((mlsize_t) ((hd) >> 10))
-#define Profinfo_hd(hd) NO_PROFINFO
-#endif /* WITH_PROFINFO */
+#define HEADER_BITS (sizeof(header_t) * CHAR_BIT)
+
+#define HEADER_TAG_BITS 8
+#define HEADER_TAG_MASK ((1ull << HEADER_TAG_BITS) - 1ull)
+
+#define HEADER_COLOR_BITS 2
+#define HEADER_COLOR_SHIFT HEADER_TAG_BITS
+#define HEADER_COLOR_MASK ((1ull << HEADER_COLOR_BITS) - 1ull)
+
+#define HEADER_WOSIZE_BITS (HEADER_BITS - HEADER_TAG_BITS - HEADER_COLOR_BITS - HEADER_RESERVED_BITS)
+#define HEADER_WOSIZE_SHIFT (HEADER_COLOR_SHIFT  + HEADER_COLOR_BITS)
+#define HEADER_WOSIZE_MASK ((1ull << HEADER_WOSIZE_BITS) - 1ull)
+
+/* Use with care: this defaults to zero if HEADER_RESERVED_BITS is
+ * zero, to avoid warnings of undefined behaviour (of a left shift by
+ * the type width). */
+#define HEADER_RESERVED_SHIFT (HEADER_RESERVED_BITS ? (HEADER_BITS - HEADER_RESERVED_BITS) : 0)
```

Comment:
Instead of this "use with care" constant, we could have two functions that morally correspond to `x << NON_RESERVED_BITS` and `x >> NON_RESERVED_BITS`. In the case where `HEaDER_RESERVED_BITS` is 0, the `x << NON_RESERVED_BITS` variant could even assert that `x == 0` (the only valid 0-bit value).

Reply:
Nice idea!

Reply:
The reason I didn't do something like this yesterday is that the compiler isn't smart enough to figure out that `(B ? (x << (64-B)) : 0)` is legit. Whereas `(B ? (x << (B ? 64-B : 0)) : 0)` is. I can write the appropriate macros, they are just ugly.